### ISSUE TITLE
Paper 

### ISSUE THREAD

**Issue by fantahund** [IzzelAliz/Arclight#47]
Can you implement the paper API?

**Comment by IzzelAliz** [comment]
I don't know exactly how much plugin requires a paper api.
This is actually planned with very low priority.

If someone see this please add an emoji here.

**Comment by fantahund** [comment]
We have a lot of plugins on our vanilla server that need the Paper API. :>

**Comment by IzzelAliz** [comment]
This is now considered in roadmap.

I'm busy schooling now, and if I have enough time I will work on it.

**Comment by DeZoWaRe** [comment]
> This is now considered in roadmap.
> 
> I'm busy schooling now, and if I have enough time I will work on it.

Paper.yml -> down

viewdistances:
      no-tick-view-distance: 8

This is what is very important for the server mod in the Paper API, with this you can save RAM and CPU.

**Comment by Mike5357** [comment]
Paper support is definitely something I'd love to see supported by Arclight

**Comment by IzzelAliz** [comment]
I wonder if users prefer Paper API support or just performance and optimize.

Please react with ❤️ if API support and 🚀 if performance.

**Comment by Hi-Im-Ina** [comment]
Hello,
At the moment the [Pixelmon mod](https://reforged.gg/) community is preparing itself for a version bump to 1.16.5 and as such we've all been looking for ways to host our servers.
The consensus seems to be SpongeForge is less preferred due to the mostly dry ecosystem of plugins, Mohist is unstable but does come with Paper benefits unlike Arclight which despite being generally considered the best solution for those looking to jump to a Spigot hybrid has always been held back compared to the alternative due to the lack of Paper API and optimizations.
Arclight has a lot to gain in terms of community support by properly implementing Paper even if just bit by bit into itself.
Hopefully this incentivizes you to consider tackling both sides of the coin.

Thank you for all the work you do in the project!

**Comment by Rasgnarok** [comment]
PixelmonMod Community Manager here. A full, clean implementation of Paper is needed for Arclight to be production-ready. Servers with a large amount of players will require the support Paper provides in API and performance. We expect that with our move to 1.16.5, you'll be seeing a lot more pressure from the server community towards Arclight's performance in comparison to projects such as Mohist or even Sponge. Should you need to discuss this further, I'm on your discord, Rasgnarok#6969.

**Comment by josephtarbit** [comment]
I think both are a good idea, performance is definitely a huge issue for large scale servers at the moment especially with chunk loading.

**Comment by IzzelAliz** [comment]
Per discussion in Discord.

> I have the interest to have Paper in Arclight. However, Paper is rapidly changing and its API set is way larger than Spigot, and they make much more changes to this game.
> Maintaining from both Spigot and Paper is too difficult for me, but if there is help from community it would be good.
> Here are some stats: Spigot has roughly 40k line patches and Paper has 118k line patches. I can made an update between major versions in about 7 days for a 40k line project. The lines in Paper should be lesser as they sometimes include api codes in server patches.
> Paper is a simple human resource problem to me, and if someone want to become a maintainer that is responsible for version upgrade, bugfix, optionally optimisations, doing paper stuff, etc please DM me.

**Comment by TransparentWhit** [comment]
At least add the adventure API plz

**Comment by qyl27** [comment]
After Paper changes more of Spigot's behaviors, we might need to re-evaluate the priority of this issue and make a decision.  
cc @IzzelAliz

**Comment by PukPukov** [comment]
Reducing paper api support is actually good thing because idiotic paper devs gives no shit to backwards compatibility. Less paper api support = less paper plugins = less pain.

**Comment by virtualspan** [comment]
This is the time to make implementing the Paper API as a **priority**, this is because as of 1.21.4, Paper is becoming a **hard fork** and distancing itself from Spigot, and it will slowly **drop Spigot plugin support** over time. This all basically means that plugins will be **forced**, to increase their workload supporting both, or _more likely_, **remove spigot support and migrate to Paper**. And obviously Paper is not going anywhere, it is the most widely used minecraft server jar, with Spigot being a much lesser and lesser priority for plugin Devs as it loses popularity. Ultimately, Arclight is going to have to support the Paper API at some point soon, unless it wants to significantly reduce plugin support.

**Comment by PukPukov** [comment]
Ah yes, the good old "big brother watching you"-style reasoning. You should give paper all rights to do everything with API without being limited by spigot not because paper good or something, no, you should do that because everyone do that and you can only cry about it.

**Comment by PukPukov** [comment]
>I wonder if users prefer Paper API support or just performance and optimize.

Paper actually does not improve performance, from my benchmarks that i've done last year Spigot and Paper had the exact same performance (Paper was even slower, but this was most likely measurement error). Only way to improve performance over Spigot was to use performance-targeted forks of Paper that could improve it twice.

**Comment by PukPukov** [comment]
Paper have not so big support — they actively using bots and their own staff members to create illusion of support (does it reminds me of something?).

![2024 12 14_08-41](https://github.com/user-attachments/assets/9703b1c9-c6f2-44d7-8ee1-13b9c2f4a1cc)
![2024 12 13_17-56](https://github.com/user-attachments/assets/11eaafd9-eaf4-4177-a1f1-5c30c6c983f8)
![2024 12 13_17-57](https://github.com/user-attachments/assets/25903098-96e5-40c4-919c-7212580dfcdd)
![2024 12 13_17-57_1](https://github.com/user-attachments/assets/80629b3f-19c8-4e78-9906-944757fa9262)
![2024 12 13_17-57_2](https://github.com/user-attachments/assets/bc0043fd-29a9-4ce8-90a7-5a01fbaf5384)
![2024 12 13_19-18](https://github.com/user-attachments/assets/d9e8b1ad-5645-48af-95f9-847da6cded83)
![2024 12 14_08-42](https://github.com/user-attachments/assets/58fdff71-5210-4fe2-ad04-37b7705e3348)
![2024 12 14_08-42_1](https://github.com/user-attachments/assets/95eff04e-f1b2-4b9e-9ad5-08d898eb18de)

And if you will try to argue with them, you will get more insults than actual arguments. Check this thread: https://www.spigotmc.org/threads/wiki-vg-is-down-—-any-alternative.242197

**Comment by PukPukov** [comment]
And the irony is that paper hardfork directly hits all forks of paper that doing better, eliminating all competitors. All of them will have to perform a huge job on transferring their own patches to new paper patch system. While paper with dozens of developers can do that, forks that have just one or two maintainers most likely will just be discontinued.

**Comment by PukPukov** [comment]
Paper again forcing developers to spend their time on adapting with a sole alternative of discontinuing their project which they could have been working on for years. Whats paper's argument on that? "ur lazy son of a bitch". As if developers owe something to paper.

**Comment by qyl27** [comment]
Yes, I stand with Paper, Spigot API is too old to face the new versions of Minecraft, and developers have suffered from Spigot for a long time.  
We will begin our Paper implementation soon, and keep Spigot API support.

**Comment by PukPukov** [comment]
>and developers have suffered from Spigot for a long time.

That does not sound very convincing without any arguments, especially when there are some quite well-reasoned posts above that says exactly opposite things.

**Comment by PukPukov** [comment]
I have to add that I am not agitating for spigot, paper also have some good things over spigot, like adventure lib with itemstacks integration and anti xray, and no need to lose it just because paper team became mad. Spigot could just pull that changes from paper, but they refused (for some reason; there could be a good reason, but I dont see any). 

What we have to do is to agree that paper have huge ideologic problems and stop being fans of somebody who completely disrespects you and your work. One of alternatives will be fork of paper that will think about backward compatibility.

It could start even from old repo, if fork authors that now forced by paper to adapt will refuse and make 1.21.4 update themselves, together.

**Comment by qyl27** [comment]
> > and developers have suffered from Spigot for a long time.
> 
> That does not sound very convincing without any arguments, especially when there are some quite well-reasoned posts above that says exactly opposite things.

Yes, something you might not know, that is for plugin developers. It doesn't matter.  
Personally, I totally disagree with your ideological accusations about the hard fork. I am focused on the technical things and second it.  
Btw, you can criticize PaperMC, but in this issue, we are talking about the implementation, and we will do it soon.

**Comment by PukPukov** [comment]
Implementation is "how", ideology is "why"... It is more important. Being fair, for Arclight hard fork matters less because it is mixin based, but paper's carte blanche on api breakages still will affect arclight later.

**Comment by qyl27** [comment]
Again, the hard fork will be positive for the server-side software, we are happy to see that and will deliver ourselves on it.

**Comment by PukPukov** [comment]
Again, "nuh uh" is not a best argument.

**Comment by IzzelAliz** [comment]
Stop making drama here.We're nothing against paper or their team. If majority of users want paper support, we'll consider it, and we are considering it.2024年12月15日 08:36，PukPukov ***@***.***>写道：
Again, "nuh uh" is not a best argument.

**Comment by virtualspan** [comment]
This is just a suggestion and I'm not sure if it's possible, but can you use [Purpur](https://purpurmc.org) instead of Paper when migrating to the Paper API? Purpur is not an API itself, as it's a fork of [Pufferfish](https://pufferfish.host/downloads) which is a fork of Paper, just adding performance improvements from Pufferfish and configuration options from Purpur on top of Paper. So basically I'd like to see the performance and configuration improvements from Purpur in Arclight, and all plugins compatible with Paper should be compatible with Purpur.

**Comment by PukPukov** [comment]
Arclight implementation have nothing to do with paper patches either, its mixin based. Whats discussed in this issue is arclight own implementation of paper api.

**Comment by virtualspan** [comment]
> Arclight implementation have nothing to do with paper patches either, its mixin based. Whats discussed in this issue is arclight own implementation of paper api.

> I wonder if users prefer Paper API support or just performance and optimize.
> 
> Please react with ❤️ if API support and 🚀 if performance.

I assumed they were implementing paper patches and could implement Purpur/Pufferfish ones from the mention of adding Paper API for performance.

**Comment by AgentMelinda** [comment]
> Arclight implementation have nothing to do with paper patches either, its mixin based. Whats discussed in this issue is arclight own implementation of paper api.

bro please be quiet and stop being so childish its not like you develop arclight XD 
both IzzelAliz and qyl27 know's what they are talking about!

**Comment by PukPukov** [comment]
> bro please be quiet and stop being so childish its not like you develop arclight XD both IzzelAliz and qyl27 know's what they are talking about!

I dont see any of them in this particular conversation.

**Comment by AgentMelinda** [comment]
> > bro please be quiet and stop being so childish its not like you develop arclight XD both IzzelAliz and qyl27 know's what they are talking about!
> 
> I dont see any of them in this particular conversation.

then you can fuck off XD its a topic about the implementation the paper API to arclight project XD you aint developing it so why talking like you above all?

**Comment by AgentMelinda** [comment]
> > bro please be quiet and stop being so childish its not like you develop arclight XD both IzzelAliz and qyl27 know's what they are talking about!
> 
> I dont see any of them in this particular conversation.

also if you are so much better then them? why not proving it by making your lame hybrid?

**Comment by PukPukov** [comment]
> then you can fuck off XD its a topic about the implementation the paper API to arclight project XD you aint developing it so why talking like you above all?

> also if you are so much better then them? why not proving it by making your lame hybrid?

I believe I only answered to posted here question about purpur patches. Sober up please.

**Comment by AgentMelinda** [comment]
> > then you can fuck off XD its a topic about the implementation the paper API to arclight project XD you aint developing it so why talking like you above all?
> 
> > also if you are so much better then them? why not proving it by making your lame hybrid?
> 
> I believe I only answered to posted here question about purpur patches. Sober up please.

after you ladies first :wink:

**Comment by PukPukov** [comment]
> after you ladies first 😉

Bruh, you really think you "won" by posting last?

**Comment by Mike5357** [comment]
This has gone completely off topic. Please stop flooding this thread with senseless bickering.

**Comment by qyl27** [comment]
No more filibuster PukPukov.  
Paper API will soon available in Arclight. We embrace the future.

**Comment by InitAuther97** [comment]
It is sure that, due to the great amount of conflicts we have seen between Arclight and other mods utilizing Mixin, a minimum set of modification to support Paper API will be made, and the optimizations are not a main goal due to their nature of destroying compatibilities.
Mixin, with its easy to use design, has been used widely in mod developments. It enables developers to easily modify the game and effectively reduce the time and energy one can put into it when implementing a certain functionality. Most people love it when making mods.
But it comes with risks. The fact is that with more and more hooks or modification made, the game inevitably becomes fragile because of the various ways we and other rely on different parts all across the game. There can and will be conflicts, but the conflicts won't stop many from using Mixin, even if they can do what they want without using Mixin. Perhaps many don't even consider that.
We have checked Moonrise, an optimization mod that ports several Paper patches to modded servers, featuring their completely overwritten chunk management system that's wanted by many, and a lot of Mixin conflicts have occurred (it even conflicts with Architectury on NeoForge and some mods we've seen that already conflicts with Arclight). 
This means these patches are, as we could see now, the least likely to be included into Arclight, because we'll have to spend time and time to take a gamble on making them compatible in modded servers, which is rare and valuable since we only have 3 active maintainers.
Still we welcome people who's interested in Arclight development, and your support will make a difference.
